tensorflow/python/ops/nn_ops.py

```diff
@@ -2193,7 +2193,31 @@ def dropout(x, keep_prob, noise_shape=None, seed=None, name=None):  # pylint: di
     if tensor_util.constant_value(keep_prob) == 1:
       return x
 
-    noise_shape = noise_shape if noise_shape is not None else array_ops.shape(x)
+    def noise_shape_func(x, noise_shape):
+      # If noise_shape is none return immediately.
+      if noise_shape is None:
+        return array_ops.shape(x)
+
+      try:
+        # Best effort to figure out the intended shape.
+        # If not possible, let the op to handle it.
+        noise_shape_ = tensor_shape.as_shape(noise_shape)
+        if (x.shape.dims is not None and
+            len(x.shape.dims) == len(noise_shape_.dims)):
+          new_dims = []
+          for i, dim in enumerate(x.shape.dims):
+            if noise_shape_.dims[i].value is None and dim.value is not None:
+              new_dims.append(dim.value)
+            else:
+              new_dims.append(noise_shape_.dims[i].value)
+          return tensor_shape.TensorShape(new_dims)
+      except Exception:
```

Comment:
What exceptions does this catch in practice?

In general we like to avoid catching exceptions this wide, and it would be good to scope it more appropriately.